ft232/mpsse.py

```diff
@@ -0,0 +1,555 @@
+            PinsValue[Pin] = self.__getPinValue(Pin, PinsConfig)
+        return PinsValue
+
+    def setClock(self, Hz, ThreePhase=False, AdaptiveClock=False):
+        # This function is inspired and partially copied from Adafruit_Python_GPIO/Adafruit_GPIO/FT232H.py on GitHub.
+        # for more information see: https://github.com/adafruit/Adafruit_Python_GPIO/blob/master/Adafruit_GPIO/FT232H.py
+
+        """Set the clock speed of the MPSSE engine. Can be any value from 92hz to 30mhz (FT232H, FT2232H, FT4232H) or 92hz to 6mhz (FT2232D) and will pick that speed or the closest speed below it. When ThreePhaseClocking is enabled clock speed drops down by 1.5 (62hz - MasterClock/3).
+        Parameter
+        ---------
+        Hz: int, 92hz to 30mhz (FT232H, FT2232H, FT4232H), 92hz to 6mhz (FT2232D)
+            Frequency in hz
+        ThreePhase: bool
+            Enables/Disables ThreePhaseClocking reduces Hz range by factor 1.5
+        AdaptiveClock: bool
+            Enables/Disables AdaptiveClocking
+        Raises
+        ------
+        ValueError:
+            Hz must be within 62 - MasterClock/3  when ThreePhaseClocking is enabled
+        ValueError:
+            Hz must be within 92 - MasterClock/2    
+        """
+        # Turn on/off adaptive clocking.
+        if AdaptiveClock:
+            self.__enableAdaptiveClock()
+        else:
+            self.__disableAdaptiveClock()
+
+        # Turn on/off three phase clock (needed for I2C).
+        # Also adjust the frequency for three-phase clocking as specified in section 2.2.4
+        # of this document:
+        #   http://www.ftdichip.com/Support/Documents/AppNotes/AN_255_USB%20to%20I2C%20Example%20using%20the%20FT232H%20and%20FT201X%20devices.pdf
+        if ThreePhase:
+            self.__enableThreePhaseClocking()
+            Hz *= 1.5
+        else:
+            self.__disableThreePhaseClocking()
+
+        # Compute divisor for requested clock.
+        # Use equation from section 3.8.1 of:
+        # http://www.ftdichip.com/Support/Documents/AppNotes/AN_108_Command_Processor_for_MPSSE_and_MCU_Host_Bus_Emulation_Modes.pdf
+        MasterClock = self.__Ft232.MPSSE_CLOCK
+        
+        if 92 <= Hz <= (MasterClock / 2):
+            if Hz < (457763) and self.__isModernDevice():
+                self.__enableClockDivide()
+                MasterClock /= 5
+            elif Hz >= (457763) and self.__isModernDevice():
+                self.__disableClockDivide()
+        elif ThreePhase:
+            raise ValueError("Hz must be within 62 - MasterClock/3  when ThreePhaseClocking is enabled")
+        else:
+            raise ValueError("Hz must be within 92 - MasterClock/2")        
+
+        Divisor = math.ceil(MasterClock / (2 * Hz) - 1) & 0xffff
+        
+        # Send command to set divisor from low and high byte values.
+        self.__setClockDivisor(Divisor)
+
+    def _write(self, Data, AsBitsBytes, WriteClkEdge, BitOrder, Tms=TMS_OFF):
```

Comment:
I don't think it makes sense to change the WriteClkEdge, BitOrder and Tms on a per write command basis. I think these are better set at class `__init__()` and used every time (Essentially like you do on the next layer up in the SPI class.)

Reply:
The MPSSE class is not indended to use as a standalone. It should only be subclassed.

Reply:
That's not a good excuse: don't write a bad interface that's not meant to be used, only to create a good interface on top of it. Just squash it down and write less code to stitch it all together. A class should still have a purpose even if it is only used privately. MPSE's purpose is very confused in that it provides SPI and GPIO functionality to other unrelated classes that probably shouldn't exist and it's existence has created a less than ideal API where the user has to create more class instances than they really need from the device class they already have.